repo/fsrepo/fsrepo.go

```diff
@@ -5,79 +5,142 @@ import (
 	repo "github.com/jbenet/go-ipfs/repo"
-	common "github.com/jbenet/go-ipfs/repo/common"
 	config "github.com/jbenet/go-ipfs/repo/config"
-	opener "github.com/jbenet/go-ipfs/repo/fsrepo/opener"
-	util "github.com/jbenet/go-ipfs/util"
+	component "github.com/jbenet/go-ipfs/repo/fsrepo/component"
+	counter "github.com/jbenet/go-ipfs/repo/fsrepo/counter"
+	dir "github.com/jbenet/go-ipfs/repo/fsrepo/dir"
+	lockfile "github.com/jbenet/go-ipfs/repo/fsrepo/lock"
+	serialize "github.com/jbenet/go-ipfs/repo/fsrepo/serialize"
 	debugerror "github.com/jbenet/go-ipfs/util/debugerror"
 )
 
 var (
-	// openerCounter prevents the fsrepo from being removed while there exist open
+
+	// packageLock must be held to while performing any operation that modifies an
+	// FSRepo's state field. This includes Init, Open, Close, and Remove.
+	packageLock sync.Mutex // protects openersCounter and lockfiles
+	// lockfiles holds references to the Closers that ensure that repos are
+	// only accessed by one process at a time.
+	lockfiles map[string]io.Closer
+	// openersCounter prevents the fsrepo from being removed while there exist open
 	// FSRepo handles. It also ensures that the Init is atomic.
 	//
 	// packageLock also protects numOpenedRepos
 	//
 	// If an operation is used when repo is Open and the operation does not
 	// change the repo's state, the package lock does not need to be acquired.
-	openerCounter *opener.Counter
+	openersCounter *counter.Openers
 )
 
 func init() {
-	openerCounter = opener.NewCounter()
+	openersCounter = counter.NewOpenersCounter()
+	lockfiles = make(map[string]io.Closer)
 }
 
-// FSRepo represents an IPFS FileSystem Repo. It is not thread-safe.
+// FSRepo represents an IPFS FileSystem Repo. It is safe for use by multiple
+// callers.
 type FSRepo struct {
-	state  state
-	path   string
-	config *config.Config
+	// state is the FSRepo's state (unopened, opened, closed)
+	state state
+	// path is the file-system path
+	path string
+	// configComponent is loaded when FSRepo is opened and kept up to date when
+	// the FSRepo is modified.
+	// TODO test
+	configComponent    component.ConfigComponent
+	datastoreComponent component.DatastoreComponent
+}
+
+type componentBuilder struct {
+	Init          component.Initializer
+	IsInitialized component.InitializationChecker
+	OpenHandler   func(*FSRepo) error
 }
 
 // At returns a handle to an FSRepo at the provided |path|.
-func At(path string) *FSRepo {
+func At(repoPath string) *FSRepo {
 	// This method must not have side-effects.
 	return &FSRepo{
-		path:  path,
+		path:  path.Clean(repoPath),
 		state: unopened, // explicitly set for clarity
 	}
 }
 
+func ConfigAt(repoPath string) (*config.Config, error) {
+
+	// packageLock must be held to ensure that the Read is atomic.
+	packageLock.Lock()
+	defer packageLock.Unlock()
+
+	configFilename, err := config.Filename(repoPath)
+	if err != nil {
+		return nil, err
+	}
+	return serialize.Load(configFilename)
+}
```

Comment:
@briantigerchow we're safe from writes from our side, but not from arbitrary writers. this might fail:

```
# write from non ipfs process
ipfs daemon & 
ipfs config foo bar & 
cat /dev/random | ~/.go-ipfs/config

# write from an ipfs client hooked up to another daemon
ipfs daemon &
ipfs config foo bar & 
ipfs config foo baz # talking to another daemon via, say an env flag.
```

I think the convention should be that only one process at a time can write to the config file, and takes a whole-repo lockfile.